### PR TITLE
Improve return codes (UnsupportedFork)

### PR DESCRIPTION
**Pull request by flcl42**
Covers corner cases revealed by the Hive

## Changes

- Return UnsupportedFork error for fcuvN if N < 3, but timestamp is Cancun or if N >=3 but timestamp belongs to any other (!=N) fork

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

### CODE REVIEW

**Review by marcindsobczak — APPROVED** [review]

**Review by benaadams — APPROVED** [review]

**Review by Demuirgos — APPROVED** [review]

File: src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs [hunk 1]
```diff
@@ -100,38 +102,69 @@ executionPayload switch
...
         {
-            error = $"PayloadAttributesV{expectedVersion} expected";
+            if (actualVersion == apiVersion && expectedVersion != apiVersion)
+            {
+                error = $"{methodName}{expectedVersion} expected";
+                return PayloadAttributesValidationResult.UnsupportedFork;
+            }
         }
-        else if (actualVersion > version)
+        else if (apiVersion == EngineApiVersions.Shanghai)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs:138]: Can we write this in a generic way, without referencing explicit constants? That was the idea, now we have `EngineApiVersions.Shanghai` and `EngineApiVersions.Cancun` referenced everywhere and same problem on next hardfork, making code more and more complex.

File: src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.Paris.cs [hunk 2]
```diff
@@ -39,10 +39,13 @@ public async Task<ResultWrapper<PayloadStatusV1>> engine_newPayloadV1(ExecutionP
...
-        if (payloadAttributes?.Validate(_specProvider, version, out string? error) == false)
+        string? error = null;
+        switch (payloadAttributes?.Validate(_specProvider, version, out error))
         {
-            if (_logger.IsWarn) _logger.Warn(error);
-            return ResultWrapper<ForkchoiceUpdatedV1Result>.Fail(error, version >= EngineApiVersions.Cancun ? ErrorCodes.UnsupportedFork : ErrorCodes.InvalidParams);
+            case PayloadAttributesValidationResult.InvalidParams:
+                return ResultWrapper<ForkchoiceUpdatedV1Result>.Fail(error!, ErrorCodes.InvalidParams);
+            case PayloadAttributesValidationResult.UnsupportedFork:
+                return ResultWrapper<ForkchoiceUpdatedV1Result>.Fail(error!, ErrorCodes.UnsupportedFork);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.Paris.cs:48]: Why we need to map same labels to same labels? Do we need `PayloadAttributesValidationResult` at all?

File: src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs (hunk 1 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs:138]: The code considers next forks, should work for any further one(for the future forks any API serves only one fork):
> if (apiVersion >= EngineApiVersions.Cancun)

File: src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.Paris.cs (hunk 2 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.Paris.cs:48]: This one is harder to describe, we need it for refactoring.
We have the similar enum `Nethermind.Merge.Plugin.Data.ValidationResult` for `ExecutionPayload` validation that also includes value that means `ErrorCode=None, Result=Invalid`. So I decided to make a separate `enum` that will merge with this one and will be different from `ErrorCodes`.

The current solution for fcu seems more correct that ones we have for get/new payload. And I expected to find more time later to refactore 3 of them to follow this one approach.

File: src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs:138]: Maybe let's have `MaxFork` constant?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs:138]: I don't get it

File: src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.Paris.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.Paris.cs:48]: Ah it maps to int's got it, though I would just `out` the error code from `Validate` method and make it bool and drop the enum.

File: src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs (hunk 1 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs:138]: It works for new forks too, rules are just different for shanghai and cancun+further_forks.

**Pull request merged**